Use a single SharedMem per stream, rather than per stream direction

kinetiknz opened the pull request:
This addresses https://github.com/mozilla/audioipc-2/issues/45 by configuring a single SharedMem per stream and using it for both directions of data transport.  Handling this is simple on the server side, as access is sequenced to the SharedMem by the interaction with the client via IPC (copy to input, call and wait on client, copy to output).  On the client side, an extra copy for the input is required for duplex streams to preserve cubeb data_callback's API requirements.

To allow Gecko to control the SharedMem region size via a pref, the SHM_AREA_SIZE constant is removed and a per-server configurable value is used.  This value is passed to the client during SharedMem setup rather than assuming a fixed size.  SharedMem setup between server and client is now handled via a new CallbackReq/CallbackResp message.

This also addresses https://github.com/mozilla/audioipc-2/issues/68 by reducing the file descriptor passing requirement to one per message.  Framed becomes obsolete as all transports are now FramedWithPlatformHandles, so frame.rs is removed.

PlatformHandle is simplified into a owning only handle for internal use. RemoteHandle is introduced to message.rs to accurately represent the OS-specific serialized state of a remote handle.

SharedMem is simplified, removing the non-owning variant.  new() now returns a single SharedMem, with handle access via make_handle() with clearer handle ownership.

ChunMinChang commented:
I've reviewed half of the patch. Will do the rest tomorrow.

ChunMinChang commented on audioipc/src/messages.rs:
```diff
@@ -175,17 +175,115 @@ fn opt_str(v: Option<Vec<u8>>) -> *mut c_char {
+    RemoteOnSender(PlatformHandleType),
+    RemoteOnReceiver(PlatformHandleType),
+    Empty,
+}
```
This looks like a great idea!

Would you mind adding more comments about how the handle is transferred, in Windows and Unix respectively? I'd hope some handle ownership (e.g., who will close the handle) can be documented here.

ChunMinChang reviewed (approved):
Sorry for the late reply.

This seems a good improvement. I'd hope some comments can be added so what RemoteHandle/PlatformHandle/SharedMem are clearer without tracing all the code.

kinetiknz commented on client/src/context.rs:
```diff
@@ -81,8 +81,9 @@ fn promote_thread(rpc: &rpc::ClientProxy<ServerMessage, ClientMessage>) {
-            rpc.call(ServerMessage::PromoteThreadToRealTime(bytes));
+            rpc.call(ServerMessage::PromoteThreadToRealTime(bytes))
+                .wait()
+                .unwrap();
```
It turns out this change wasn't needed, so I dropped it.

kinetiknz commented:
I had set these changes aside while working on the core IPC rework for macOS.  That work depends on some of the refactoring in this PR, so I'd like to merge this to simplify merging the incoming core IPC/macOS changes.

- Addressed review comments
- Renamed RemoteHandle to SerializableHandle and documented lifecycle during message send/receive
- Clean up mixed use of "fd" and "handle" terminology in OS-agnostic code, now using "handle" consistently
- Merged fd_passing.rs and handle_passing.rs into new framing.rs, reducing code duplication between Windows/Unix

Given the additional changes since the first review, r? @ChunMinChang again please.

ChunMinChang commented:
Sorry for the delay. I am working on the review right now. Will finish this ASAP.

ChunMinChang commented on client/src/stream.rs:
```diff
@@ -152,6 +193,20 @@ impl rpc::Server for CallbackServer {
+                self.shm = Some(shm);
+
+                self.duplex_input = if let StreamDirection::Duplex = self.dir {
+                    Some(Vec::with_capacity(shm_area_size))
```
How do we know `shm_area_size` is larger than `nframes as usize * input_frame_size`?

kinetiknz replied:
Unfortunately we don't have a way to precisely determine the required shmem size right now, so we choose a sufficiently large over-estimate instead (which will be specified by a Gecko pref).  The ideal solution to this is to extend cubeb's API with a way to query a configured stream's required buffer size before starting it - that will also require auditing each backend to ensure the reported buffer size is fixed (some backends may grow the buffer in response to latency, for instance).  I filed PR https://github.com/mozilla/audioipc-2/issues/124 to track this.

kinetiknz replied:
Answered here: https://github.com/mozilla/audioipc-2/pull/121#discussion_r683090394

ChunMinChang reviewed (commented):
Overall looks good!

Just one thing I'd like to check before landing this. It seems now *shmem* is a shared memory for audio input data and audio output buffer. Is there any hard requirement for the *shmem* size (e.g., it need to be larger than or equal to max(input-data-bytes-size, output-buffer-bytes-size))? Or we those data can be chunked automatically by our codec?

kinetiknz commented on client/src/stream.rs:
```diff
@@ -83,34 +93,65 @@ impl rpc::Server for CallbackServer {
+                    // Input and output reuse the same shmem backing.  Unfortunately, cubeb's data_callback isn't
+                    // specified in such a way that would require the callee to consume all of the input before
+                    // writing to the output (i.e., it is passed as two pointers that aren't expected to alias).
+                    // That means we need to copy the input here.
```
Your understanding is correct in the first diagram.  Splitting the shmem is a trade-off, it requires 2x the size but avoids a copy - since we're currently over-estimating the required size (see my reply to your next question), a copy seems like the cheaper option than doubling the shmem requirement for each stream.  We can revisit once we have precisely sized shmems; there are other improvements that can be made in this area once that's addressed.

kinetiknz commented on server/src/server.rs:
```diff
@@ -681,18 +672,14 @@ impl CubebServer {
-            SharedMem::new(&format!("{}-input", shm_id), audioipc::SHM_AREA_SIZE)?;
-        let (output_shm, output_file) =
-            SharedMem::new(&format!("{}-output", shm_id), audioipc::SHM_AREA_SIZE)?;
+        let shm = SharedMem::new(&get_shm_id(), self.shm_area_size)?;
```
Unfortunately we don't have a way to precisely determine the required shmem size right now, so we choose a sufficiently large over-estimate instead (which will be specified by a Gecko pref).  The ideal solution to this is to extend cubeb's API with a way to query a configured stream's required buffer size before starting it - that will also require auditing each backend to ensure the reported buffer size is fixed (some backends may grow the buffer in response to latency, for instance).  I filed PR https://github.com/mozilla/audioipc-2/issues/124 to track this.

kinetiknz commented on audioipc/src/messages.rs:
```diff
@@ -285,14 +285,141 @@ pub enum DeviceCollectionResp {
+    type Value = SerializableHandle;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        formatter.write_str("an integer between -2^63 and 2^63")
```
We don't have access to the value failed to deserialize from here.  This error shouldn't ever be triggered, and I can't think of anything useful to report here if it was.
